client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java

```diff
@@ -756,15 +779,19 @@ public static ContentType createContentType(final XContentType xContentType) {
      * a {@link Request} and adds the parameters to it directly.
      */
     static class Params {
-        private final Request request;
+        private final Map<String, String> requestParamsMap;
+
+        Params() {
+            this.requestParamsMap = new HashMap<>();
+        }
 
-        Params(Request request) {
-            this.request = request;
+        Map<String, String> getRequestParams(){
+            return this.requestParamsMap;
         }
 
         Params putParam(String name, String value) {
             if (Strings.hasLength(value)) {
-                request.addParameter(name, value);
+                requestParamsMap.put(name, value);
```

Comment:
with this change we lose two checks that we were previously performing by calling Request#addParameter: 
- check that the key is not null
- check that the param has not already been set (it would get overridden by default, but we prefer to throw error)

I think that we would want to keep these two checks.

Reply:
@javanna what error should i throw? i think that check would move to the copy method of Request class
```suggestion
                   public void setParameters(Map<String, String> paramSource){
        paramSource.keySet().forEach(key -> {
            if(this.parameters.get(key) != null){
                throw new Error(String.format("%s is already set", key));
            }
        });
        this.parameters.putAll(paramSource);
    }
```
does this look good?

Reply:
can you throw the same error that we throw in the Request class please? IllegalArgumentException?